File: number-theory/primality-test.py
import sys
import math


try:
    sys.stdin = open(sys.path[0] + '\\input.txt', 'r')
    sys.stdout = open(sys.path[0] + '\\output.txt', 'w')
except FileNotFoundError:
    pass


def primality_naive(n):
    # O(n) linear
    if n <= 1:
        return False

    for i in range(2, n):
        if n % i == 0:
            return False
    return True


def primality_logn(n):
    # O(log n) logarithmic
    if n <= 1:
        return False

    if n == 2:
        return True

    for i in range(3, int(n ** 0.5) + 1, 2):
        if n % i == 0:
            return False
    return True


print(primality_naive(10000))
print(primality_logn(3))

File: number-theory/primality-test.cpp
#include <bits/stdc++.h>

#define MAX 1000005
#define ll long long
#define watch(x) cout << (#x) << " is " << (x) << endl

using namespace std;


bool primality_naive(ll num) {
    if (num <= 1) return false;

    for (int i = 2; i < num; i++) {
        if (num % i == 0)
            return false;
    }

    return true;
}


bool primality_logn(ll num) {
    if (num <= 1) return false;
    if (num == 2) return true;

    for (int i = 3; i <= sqrt(num); i++) {
        if (num % i == 0)
            return false;
    }

    return true;
}


int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(NULL);

    #ifndef ONLINE_JUDGE
    freopen("input.txt", "r", stdin);
    freopen("output.txt", "w", stdout);
    #endif

    cout << primality_naive(10001) << endl;
    cout << primality_logn(3) << endl;

    return 0;
}